File: app/src/main/java/org/mifos/mobilebanking/api/local/PreferencesHelper.java
package org.mifos.mobilebanking.api.local;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.text.TextUtils;

import org.mifos.mobilebanking.api.BaseURL;
import org.mifos.mobilebanking.api.SelfServiceInterceptor;
import org.mifos.mobilebanking.injection.ApplicationContext;

import java.util.Map;

import javax.inject.Inject;
import javax.inject.Singleton;

/**
 * @author Vishwajeet
 * @since 07/06/16
 */

@Singleton
public class PreferencesHelper {
    private static final String USER_ID = "preferences_user_id";
    private static final String TOKEN = "preferences_token";
    private static final String CLIENT_ID = "preferences_client";
    private static final String OFFICE_NAME = "preferences_office_name";
    private static final String USER_NAME = "preferences_user_name";
    private static final String PASSCODE = "preferences_passcode";
    private static final String OVERVIEW_STATE = "preferences_overview_state";
    private static final String SENT_TOKEN_TO_SERVER = "sentTokenToServer";
    private static final String GCM_TOKEN = "gcm_token";
    private static final String TENANT = "preferences_base_tenant";
    private static final String BASE_URL = "preferences_base_url_key";
    private static final String PROFILE_IMAGE = "preferences_profile_image";
    public static final String CLIENT_NAME = "client_name";

    private SharedPreferences sharedPreferences;

    @Inject
    public PreferencesHelper(@ApplicationContext Context context) {
        this.sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
    }

    public void clear() {
        SharedPreferences.Editor editor = sharedPreferences.edit();
        //prevent deletion of url and tenant
        for (Map.Entry<String, ?> entry: sharedPreferences.getAll().entrySet()) {
            if (!(entry.getKey().equals(BASE_URL) || entry.getKey().equals(TENANT))) {
                editor.remove(entry.getKey());
            }
        }
        editor.apply();
    }

    public int getInt(String preferenceKey, int preferenceDefaultValue) {
        return sharedPreferences.getInt(preferenceKey, preferenceDefaultValue);
    }

    public void putInt(String preferenceKey, int preferenceValue) {
        sharedPreferences.edit().putInt(preferenceKey, preferenceValue).apply();
    }

    public long getLong(String preferenceKey, long preferenceDefaultValue) {
        return sharedPreferences.getLong(preferenceKey, preferenceDefaultValue);
    }

    public void putLong(String preferenceKey, long preferenceValue) {
        sharedPreferences.edit().putLong(preferenceKey, preferenceValue).apply();
    }

    public String getString(String preferenceKey, String preferenceDefaultValue) {
        return sharedPreferences.getString(preferenceKey, preferenceDefaultValue);
    }

    public void putString(String preferenceKey, String preferenceValue) {
        sharedPreferences.edit().putString(preferenceKey, preferenceValue).apply();
    }

    public void putBoolean(String preferenceKey, boolean preferenceValue) {
        sharedPreferences.edit().putBoolean(preferenceKey, preferenceValue).apply();
    }

    public boolean getBoolean(String preferenceKey, boolean preferenceDefaultValue) {
        return sharedPreferences.getBoolean(preferenceKey, preferenceDefaultValue);
    }

    public void saveToken(String token) {
        putString(TOKEN, token);
    }

    public void clearToken() {
        putString(TOKEN, "");
    }

    public String getToken() {
        return getString(TOKEN, "");
    }

    public boolean isAuthenticated() {
        return !TextUtils.isEmpty(getToken());
    }

    public long getUserId() {
        return getLong(USER_ID, -1);
    }

    public void setUserId(long id) {
        putLong(USER_ID, id);
    }

    public String getTenant() {
        return getString(TENANT, SelfServiceInterceptor.DEFAULT_TENANT);
    }

    public void setPasscode(String passcode) {
        putString(PASSCODE, passcode);
    }

    public String getPasscode() {
        return getString(PASSCODE, "");
    }

    public void setClientId(long clientId) {
        putLong(CLIENT_ID, clientId);
    }

    public long getClientId() {
        return getLong(CLIENT_ID, -1);
    }

    public void setOfficeName(String officeName) {
        putString(OFFICE_NAME, officeName);
    }

    public String getUserName() {
        return getString(USER_NAME, "");
    }

    public void setUserName(String userName) {
        putString(USER_NAME, userName);
    }

    public String getClientName() {
        return getString(CLIENT_NAME, "");
    }

    public void setClientName(String clientName) {
        putString(CLIENT_NAME, clientName);
    }

    public String getOfficeName() {
        return getString(OFFICE_NAME, "");
    }

    public void setOverviewState(boolean state) {
        putBoolean(OVERVIEW_STATE, state);
    }

    public boolean overviewState() {
        return getBoolean(OVERVIEW_STATE, true);
    }

    public void saveGcmToken(String token) {
        putString(GCM_TOKEN, token);
    }

    public String getUserProfileImage() {
        return getString(PROFILE_IMAGE, null);
    }

    public void setUserProfileImage(String image) {
        putString(PROFILE_IMAGE, image);
    }

    public String getGcmToken() {
        return getString(GCM_TOKEN, "");
    }

    public void setSentTokenToServer(boolean sentTokenToServer) {
        putBoolean(SENT_TOKEN_TO_SERVER, sentTokenToServer);
    }

    public boolean sentTokenToServerState() {
        return getBoolean(SENT_TOKEN_TO_SERVER, false);
    }

    public void updateConfiguration(String baseUrl, String tenant) {
        sharedPreferences.edit()
                .putString(BASE_URL, baseUrl)
                .putString(TENANT, tenant)
                .apply();
    }

    public String getBaseUrl() {
        return getString(BASE_URL, new BaseURL().getDefaultBaseUrl());
    }
}

File: app/src/main/java/org/mifos/mobile/api/local/PreferencesHelper.kt
package org.mifos.mobile.api.local

import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import android.text.TextUtils
import org.mifos.mobile.api.BaseURL
import org.mifos.mobile.api.SelfServiceInterceptor
import org.mifos.mobile.injection.ApplicationContext
import javax.inject.Inject
import javax.inject.Singleton

/**
 * @author Vishwajeet
 * @since 07/06/16
 */
@Singleton
class PreferencesHelper @Inject constructor(@ApplicationContext context: Context?) {
    private val sharedPreferences: SharedPreferences? = PreferenceManager.getDefaultSharedPreferences(context)
    fun clear() {
        val editor = sharedPreferences?.edit()
        //prevent deletion of url and tenant
        if (sharedPreferences != null)
            for ((key) in sharedPreferences.all) {
                if (!(key == BASE_URL || key == TENANT)) {
                    editor?.remove(key)
                }
            }
        editor?.apply()
    }

    fun getInt(preferenceKey: String?, preferenceDefaultValue: Int): Int? {
        return sharedPreferences?.getInt(preferenceKey, preferenceDefaultValue)
    }

    fun putInt(preferenceKey: String?, preferenceValue: Int) {
        sharedPreferences?.edit()?.putInt(preferenceKey, preferenceValue)?.apply()
    }

    fun getLong(preferenceKey: String?, preferenceDefaultValue: Long): Long? {
        return sharedPreferences?.getLong(preferenceKey, preferenceDefaultValue)
    }

    fun putLong(preferenceKey: String?, preferenceValue: Long) {
        sharedPreferences?.edit()?.putLong(preferenceKey, preferenceValue)?.apply()
    }

    fun getString(preferenceKey: String?, preferenceDefaultValue: String?): String? {
        return sharedPreferences?.getString(preferenceKey, preferenceDefaultValue)
    }

    private fun putString(preferenceKey: String?, preferenceValue: String?) {
        sharedPreferences?.edit()?.putString(preferenceKey, preferenceValue)?.apply()
    }

    fun putBoolean(preferenceKey: String?, preferenceValue: Boolean) {
        sharedPreferences?.edit()?.putBoolean(preferenceKey, preferenceValue)?.apply()
    }

    fun getBoolean(preferenceKey: String?, preferenceDefaultValue: Boolean): Boolean? {
        return sharedPreferences?.getBoolean(preferenceKey, preferenceDefaultValue)
    }

    fun saveToken(token: String?) {
        putString(TOKEN, token)
    }

    fun clearToken() {
        putString(TOKEN, "")
    }

    val token: String?
        get() = getString(TOKEN, "")
    val isAuthenticated: Boolean?
        get() = !TextUtils.isEmpty(token)
    var userId: Long?
        get() = getLong(USER_ID, -1)
        set(id) {
            if (id != null)
                putLong(USER_ID, id)
        }
    val tenant: String?
        get() = getString(TENANT, SelfServiceInterceptor.DEFAULT_TENANT)
    var passcode: String?
        get() = getString(PASSCODE, "")
        set(passcode) {
            putString(PASSCODE, passcode)
        }
    var clientId: Long?
        get() = getLong(CLIENT_ID, -1)
        set(clientId) {
            if (clientId != null)
                putLong(CLIENT_ID, clientId)
        }
    var userName: String?
        get() = getString(USER_NAME, "")
        set(userName) {
            putString(USER_NAME, userName)
        }
    var clientName: String?
        get() = getString(CLIENT_NAME, "")
        set(clientName) {
            putString(CLIENT_NAME, clientName)
        }
    var officeName: String?
        get() = getString(OFFICE_NAME, "")
        set(officeName) {
            putString(OFFICE_NAME, officeName)
        }

    fun setOverviewState(state: Boolean) {
        putBoolean(OVERVIEW_STATE, state)
    }

    fun overviewState(): Boolean? {
        return getBoolean(OVERVIEW_STATE, true)
    }

    fun saveGcmToken(token: String?) {
        putString(GCM_TOKEN, token)
    }

    var userProfileImage: String?
        get() = getString(PROFILE_IMAGE, null)
        set(image) {
            putString(PROFILE_IMAGE, image)
        }
    val gcmToken: String?
        get() = getString(GCM_TOKEN, "")

    fun setSentTokenToServer(sentTokenToServer: Boolean) {
        putBoolean(SENT_TOKEN_TO_SERVER, sentTokenToServer)
    }

    fun sentTokenToServerState(): Boolean? {
        return getBoolean(SENT_TOKEN_TO_SERVER, false)
    }

    fun updateConfiguration(baseUrl: String?, tenant: String?) {
        sharedPreferences?.edit()
                ?.putString(BASE_URL, baseUrl)
                ?.putString(TENANT, tenant)
                ?.apply()
    }

    val baseUrl: String?
        get() = getString(BASE_URL, BaseURL().defaultBaseUrl)

    companion object {
        private const val USER_ID = "preferences_user_id"
        private const val TOKEN = "preferences_token"
        private const val CLIENT_ID = "preferences_client"
        private const val OFFICE_NAME = "preferences_office_name"
        private const val USER_NAME = "preferences_user_name"
        private const val PASSCODE = "preferences_passcode"
        private const val OVERVIEW_STATE = "preferences_overview_state"
        private const val SENT_TOKEN_TO_SERVER = "sentTokenToServer"
        private const val GCM_TOKEN = "gcm_token"
        private const val TENANT = "preferences_base_tenant"
        private const val BASE_URL = "preferences_base_url_key"
        private const val PROFILE_IMAGE = "preferences_profile_image"
        const val CLIENT_NAME = "client_name"
    }

}